slick/src/main/scala/slick/jdbc/DriverDataSource.scala

```diff
@@ -46,42 +46,45 @@ class DriverDataSource(
   // State that gets initialized by `init`
   @volatile private[this] var registered: Boolean = false
   @volatile private[this] var initialized = false
-  private[this] var driver: Driver = _
-  private[this] var connectionProps: Properties = _
+  @volatile private[this] var driver: Driver = _
+  @volatile private[this] var connectionProps: Properties = _
```

Comment:
Hmm, this is the reasoning:

Two threads. Thread0 and Thread1.

T_0: Thread0 initializes non-volatile `driver` to null and volatile `initialized` to false at class init

T_1: Both Thread0 and Thread1 calls getConnection and is just about to enter init

T_2: Thread1 overtakes execution of Thread0, reaches init, enters synchronized block, initializes everything, writes a value to `driver` (within the synchronized block so it's written through to main memory) and sets `initialized` to true

T_3: Thread0 enters init, reads volatile `initialized` with latest value and returns immediately since `initialized` is true

T_4: Thread0 exits init and reads `driver`. `driver` is not a volatile variable and is not read within a synchronized block so the read is not a read-through. `driver` happens to be cached in L2 cache with the original value of null. A NPE results.